azkaban-common/src/main/java/azkaban/trigger/TriggerManager.java

```diff
@@ -176,6 +179,14 @@ public List<Trigger> getTriggers() {
     return new ArrayList<>(triggerIdMap.values());
   }
 
+  // get a list of removed triggers and clear the list
+  @Override
+  public List<Integer> getRemovedTriggerIds() {
+    List<Integer> removedTriggerIdsCopy = new ArrayList<>(removedTriggerIds);
+    removedTriggerIds.clear();
+    return removedTriggerIdsCopy;
+  }
```

Comment:
question: after clearing the list, if the later clean up logic got exception or broken, will those to-be-removed triggers got captured again in the future?

Reply:
No it would just be removed in memory, as the exception only possibly happen at DB layer. So for this time, we would stop looping this trigger but if webserver restart, this logic would happen again. "RemovedTriggerId" also is only valid in memory layer only, as it is used to clean scheduleUI cache.

Reply:
So is it true that, though removed from memory at this time, if they are not really correctly cleaned due to exception, they will still be captured and cleaned in the next round?

Reply:
yes, the "next time" means when web server restarts. Those triggers will be re-detected and try to cleanup again.
As for "this time". During this session, we don't have reload logic from DB. We only can do that if user are adding/updating schedules. So they're gone in the code this time.